pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1688,13 +1690,22 @@ private void assignInstances(TableConfig tableConfig, boolean override) {
       }
     }
 
+    boolean isTableInGroup = TableConfigUtils.isTableInGroup(tableConfig);
```

Comment:
What are the scenarios that we support ?

**Adding a new table to TableGroup that pre-exists with 1 or more tables.** 

In such scenario, the replica group based InstancePartitions for the group should pre-exist and when `addTable` calls comes in `PinotHelixResourceManager` and internally calls `assignInstances`, no new instance assignment should happen and the group's existing `InstancePartitions` should be retrieved and persisted in ZK for the table

**Creating a new TableGroup with new tables.** 

- When the group is created, we don't do instance assignment and just do metadata change w.r.t that group exists without any tables and no instances. 
- When the first new table is created and added to the group, the `addTable -> assignInstances` call should recognize that this table is part of a group for which instance assignment is yet to happen. So we call the driver to assign instances for the group.
- When the 2nd, 3rd ... table is created and added to the group, the `addTable -> assignInstances` call should recognize that this table is part of a group for which instance assignment is already done.

The other way to do is that when TableGroup is created, it triggers InstanceAssignment by calling the driver. So when 1st, 2nd, 3rd table is added, it finds it is member of an existing group with existing assignment so it is pretty much NO-OP

It seems like code at line 1701 is assuming that InstancePartitions for a TableGroup should exist. But I don't see instance assignment being done during creation of TableGroup. So how do we handle the case of 1st table ?

Reply:
Furthermore 

**Creating a new TableGroup out of N existing tables**

Option 1 - 

 - Create TableGroup and assign instances so InstancePartitions for the group are computed and persisted
 - Trigger rebalance on each of the N tables with reassignInstances set to true
 - The rebalance code will call the instance assignment driver and it should return the existing InstancePartitions for the group.
 - Segments are then assigned based on the returned InstancePartitions

Option 2 - 

- Create TableGroup but don't do instance assignment.
- Trigger rebalance on each of the N tables with reassignInstances set to true.
- The rebalance code will call the instance assignment driver and it should detect that InstancePartitions don't exit for the group, so it assigns instance which are then used for segment assignment. 

It would be good to get clarity (and add brief details to code comments/docs as well) on how such scenarios are handled.

Reply:
> **Adding a new table to TableGroup that pre-exists with 1 or more tables.**

Yes this is supported. To do this one can edit the table-config for the corresponding table and set the `groupName` field. You will have to initiate a rebalance on the table to ensure data is moved around appropriately for that table.

> The other way to do is that when TableGroup is created, it triggers InstanceAssignment by calling the driver.

Yup this is how I am doing it right now. You can refer to the createTableGroup API in PinotTableGroupRestletResource. When the table-group is created, we perform instance assignment for the same and persist it in ZK for future re-use.

When we need to update the instance assignment for the group (due to addition/removal of instances), we can initiate a "Re-compute Table Group" operation from the UI and will then have to call Rebalance for each of the tables. The "Re-compute" operation will do the instance assignment for the group again and update the same in ZK. When we initiate rebalance for a table in the group, data will move around for that table based on the newly updated instance partitions for the group.

Later, I am also planning to add a "Re-balance Table Group" option in the UI which should move the data for all the tables in the group.

Reply:
> Creating a new TableGroup out of N existing tables

Yes current approach is based on option-1.

> Adding an existing table to an existing TableGroup
>
> This is like steps 2,3,4 of option 1 mentioned above. Correct ?

Yes that's correct.

I'll update the code and doc with more details on this.